scripts/ci/pre_commit/pre_commit_check_watcher_in_examples.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import sys
+from pathlib import Path
+from typing import List
+
+from rich.console import Console
+
+if __name__ not in ("__main__", "__mp_main__"):
+    raise SystemExit(
+        "This file is intended to be executed as an executable program. You cannot use it as a module."
+        f"To run this script, run the ./{__file__} command [FILE] ..."
+    )
+
+
+console = Console(color_system="standard", width=200)
+
+errors: List[str] = []
+
+WATCHER_APPEND_INSTRUCTION = "list(dag.tasks) >> watcher()"
+
+PYTEST_FUNCTION = """
+def test_run():
```

Comment:
I see that below this is being copy/pasted into each example dag file. What if this function needs to be modified or fixed in the future? Would that involve patching every single example dag file?

Reply:
Unfortunately yes. Although in this case it isn't complicated to replace few lines - one file or 100 it doesn't matter for your awk/sed script ;). It's bit of the trade off - we could also leave example dags without any pytest method and generate tests dynamically but by adding this method it's easier to run just one system tests.

Reply:
> one file or 100 it doesn't matter for your awk/sed script

Hmm, this is not the kind of precedent I want to be setting :smile: It's still quite finicky and error prone.

You can clean this up quite a bit using python itself. The `test_run` method can be defined in one location in a separate module and then just imported into each example dag file.

I've mocked up an example, see the attached image below:

![Screenshot from 2022-03-17 11-54-21](https://user-images.githubusercontent.com/65743084/158875769-241827a0-f4e1-4bd1-8fdd-969db8ea55ab.png)

Reply:
That's a nice idea. If pytest still discovers it this way, I lke it a lot better. also you can even connect it with local import to make it even more localized (same reason as in the other comment , where we care about "example_dag" being still an example dag despite also being a self-contained test. Also it nicely solves the problem which I thought about - what happens if we have two dags in an example. They would have to have slightiy different method ( dag1.clear() dag1.run() vs dag2.clear(() dag2.run()). Passing dag as parameter solves it nicely.

I'd be for something like this:

```
     # Needed to run the example dag as system test <link to the docs>
     from tests.airflow.proivders.util import get_test_run
     test_run = get_test_run(dag)
```

Reply:
> Also it nicely solves the problem which I thought about - what happens if we have two dags in an example.

Yupp, good catch! I meant to add this to my original post but totally forgot. It makes it much more configurable this way.

Reply:
Updated the code with the test method proposed by @o-nikolas (just added  # noqa: E402 since we're importing not on the top of the file)